bayes_kit/rhat.py

```diff
@@ -1,11 +1,88 @@
 import numpy as np
 from numpy.typing import NDArray, ArrayLike
+import scipy as sp
 
 FloatType = np.float64
 VectorType = NDArray[FloatType]
 SeqType = ArrayLike
 
 
+def split_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Return a list of the input chains split in half.  The result will be
+    a list twice as long as the input.  For example,
+    ```
+    >>> split_chains([[1, 2, 3, 4], [5, 6, 7]])
+    [[1, 2], [3, 4], [5, 6], [7]]
+    ```
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    List of input chains split in half.
+    """
+    return [list(arr) for chain in chains for arr in np.array_split(chain, 2)]
+
+
+def rank_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Returns a copy of the included Markov chains transformed with
+    ranks normalized to [0, 1] and an offset inverse CDF. The ranks
+    are ascending and start with 1 for the smallest value.
+
+    Parameters:
+    chains: list of univariate Markov chains
+
+    Returns:
+    List of chains with values replaced by transformed ranks.
+    """
+    if len(chains) == 0:
+        return chains
+    flattened = np.concatenate(chains)
+    ranks = flattened.argsort().argsort() + 1
+    reshaped_arrays = []
+    current_index = 0
+    for array in chains:
+        size = len(array)
+        reshaped_arrays.append(ranks[current_index : current_index + size])
+        current_index += size
+    return reshaped_arrays
+
+
+def rank_normalize_chains(chains: list[SeqType]) -> list[SeqType]:
+    """Return the rank-normalized version of the input chains.
+
+    The rank-normalized value for element `j` of list `i` is
+    ```
+    inverse_Phi((rank[i][j] - 3/8) / (size(chains) - 1/4),
+    ```
+    where `inv_Phi` is the inverse cumulative distribution function for
+    the standard normal distribution and
+    ```
+    rank[i][j] = rank_chains(chains)[i][j]
+    ```
+    is the rank of element `i` in chain `j` and `size(chains)` is the
+    total number of elements in the chains.
+
+    For a specification of ranking, see :func:`rank_chains`.
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    List of chains with values replaced by rank-normalized values.
+    """
+    S = sum([len(chain) for chain in chains])
+    result = chains
```

Comment:
> There's no need for a full deep-copy here since you aren't actually preserving any of the pre-exisitng values. Could either preallocate len(ranked_chains) arrays?

Could you maybe suggest how I do that?  I don't know how to preallocate or what you're suggesting I do with it.

I also don't know what the `[[]] * 5` example is supposed to be warning me about.

I do understand copying vs. reference, just not how it applies in Python.

Reply:
> Could you maybe suggest how I [preallocate arrays]?

Performance-wise, it may not make too much of a difference to pre-allocate vs. just doing this:
```python
result = []

for i, chain_i in enumerate(ranked_chains):
    result.append([]) # add a new list to result variable
    for j, rank_ij in enumerate(chain_i):
        val = sp.stats.norm.ppf((rank_ij - 0.325) / (S - 0.25))
        result[i].append(val)
return result
```
assuming the `i`s and `j`s are well-behaved and always start at 0 & increment, which they should here.

It depends how much data there is whether constantly allocating new elements will actually be noticeable. (Python list allocations start with 4 by default and double every time you reach the end of allocated space, but they're pretty quick about it, so the calls to `append` may not actually hurt you noticeably).

Since the chains may have different lengths, I think the best you could do would be:

```python
result = list(None for n in range(len(ranked_chains)))
for i, chain_i in enumerate(ranked_chains):
    result[i] = list(None for n in range(len(chain_i)))
    for j, rank_ij in enumerate(chain_i):
        val = sp.stats.norm.ppf((rank_ij - 0.325) / (S - 0.25))
        result[i][j] = val
return result
```
It is also possible to do this as a one-liner:
```python
result = list(list(None for x in range(len(chain_i))) for chain_i in ranked_chains)
```
but I can't imagine there would be a performance improvement that would justify writing something so hard to parse. You'd be better off just using `numpy` with a predefined size if it actually mattered:

```python
result = np.empty((n,m), dtype='uint16')
```

assuming you have a square array and don't mind getting uninitialized memory (`np.empty`); can use `np.zeros` the same way if you need it zeroed out.

---
Regarding the `[[]] * 5` example--Python has a builtin to allocate lists of known length:
```python
In [14]: foo = [5] * 10

In [15]: foo
Out[15]: [5, 5, 5, 5, 5, 5, 5, 5, 5, 5]
```
The problem is if you preallocate a list of objects this way--including a list of empty lists--it will allocate each element of the list you're constructing to point to the *same* instance of the object. So if you try to build a list of lists as `bar = [[]] * 5` thinking to get 5 empty lists, you actually get 5 pointers to the *same* empty list, with predictably hilarious results when you try to operate on them individually and wind up modifying them all, every time. (I honestly don't know why that functionality is even still in the language spec.)

Reply:
OK, used a comprehension.